File: frontend/src/types/rag.ts
export type DocStatus = 'pending' | 'processing' | 'ready' | 'failed'

export interface Document {
  id: string
  filename: string
  file_type: string
  processing_status: DocStatus
  page_count: number | null
  created_at: string
}

export interface GeneratedOption {
  option_key: 'A' | 'B' | 'C' | 'D'
  text: string
  is_correct: boolean
}

export interface GeneratedQuestion {
  text: string
  explanation: string
  options: GeneratedOption[]
}

export interface GenerateQuestionsPayload {
  document_id: string
  subject_id: string
  topic_id?: string | null
  count: number
  difficulty: 'easy' | 'medium' | 'hard'
}

export interface SaveQuestionsPayload {
  subject_id: string
  topic_id?: string | null
  questions: GeneratedQuestion[]
}

export interface SaveQuestionsResult {
  saved_count: number
  question_ids: string[]
}

export interface ChunkResult {
  content: string
  page_number: number | null
  filename: string
  similarity_score: number
}

export interface SearchResult {
  chunks: ChunkResult[]
}

File: backend/app/schemas/rag.py
import enum
from datetime import datetime
from typing import Literal
from uuid import UUID

from pydantic import BaseModel


class DocStatus(str, enum.Enum):
    pending = "pending"
    processing = "processing"
    ready = "ready"
    failed = "failed"


class DocumentOut(BaseModel):
    id: UUID
    filename: str
    file_type: str
    processing_status: DocStatus
    page_count: int | None = None
    created_at: datetime

    class Config:
        from_attributes = True


class GenerateQuestionsIn(BaseModel):
    document_id: UUID
    subject_id: UUID
    topic_id: UUID | None = None
    count: int
    difficulty: Literal["easy", "medium", "hard"]


class GeneratedOption(BaseModel):
    option_key: Literal["A", "B", "C", "D"]
    text: str
    is_correct: bool


class GeneratedQuestion(BaseModel):
    text: str
    explanation: str
    options: list[GeneratedOption]


class SaveQuestionsIn(BaseModel):
    subject_id: UUID
    topic_id: UUID | None = None
    questions: list[GeneratedQuestion]


class SaveQuestionsOut(BaseModel):
    saved_count: int
    question_ids: list[UUID]


class ChunkResult(BaseModel):
    content: str
    page_number: int | None = None
    filename: str
    similarity_score: float


class SearchResult(BaseModel):
    chunks: list[ChunkResult]